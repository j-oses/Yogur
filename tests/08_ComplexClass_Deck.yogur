// Una clase que representa una baraja de cartas.
// Para probar clases anidadas, llamadas a funciones de clase...

// The RNG seed
var seed: Int = 23456

// A mod function
def mod(a: Int, b: Int) -> m: Int {
	m = a - (a / b) * b
}

// A function which generates a pseudorandom value using a linear generator
def random() -> n: Int {
    seed = mod(seed * 15245 + 3784, 0xffff)
    n = seed
}

// Class representing a card of a small deck
class Card {
	var suit: Int	// 0 to 3
	var num: Int	// 1 to 5

	def copy(another: Card) {
		suit = another.suit
		num = another.num
	}
}

// Class representing a deck of 20 cards
class Deck {
	var cards: Card[20]
	var currentN: Int

	def generateDeck() {
		for suit in 0 to 3 {
			for num in 1 to 5 {
				var ind: Int = suit * 5 + num - 1
				cards[ind].suit = suit
				cards[ind].num = num
			}
		}
		currentN = 20
	}

	// Fisher-Yates algorithm
	def shuffle() {
		for i in 0 to (currentN - 2) {
			// i <= j < n
			var j: Int = mod(random(), (currentN - i)) + i
			swapAt(i, j)
		}
	}

	def swapAt(i: Int, j: Int) {
		var aux: Card
		aux.copy(cards[i])
		cards[i].copy(cards[j])
		cards[j].copy(aux)
	}

	def getTopCard(result: Card) -> success: Bool {
		if (currentN > 0) {
			currentN = currentN - 1
			result.copy(cards[currentN])
			success = true
		} else {
			success = false
		}
	}
}

var deck: Deck
deck.generateDeck()
deck.shuffle()
var card: Card
var success: Bool = deck.getTopCard(card)