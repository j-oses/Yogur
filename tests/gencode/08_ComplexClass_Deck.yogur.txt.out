Run successfully completed
Final P-machine state:
PC = 253   MP = 0   SP = 44   EP = -1   NP = 100   code size = 254
STORE = 
(0,Int 47714)
(1,Int 0)
(2,Int 5)
(3,Int 0)
(4,Int 2)
(5,Int 1)
(6,Int 3)
(7,Int 3)
(8,Int 2)
(9,Int 3)
(10,Int 1)
(11,Int 3)
(12,Int 5)
(13,Int 2)
(14,Int 4)
(15,Int 2)
(16,Int 1)
(17,Int 1)
(18,Int 2)
(19,Int 3)
(20,Int 4)
(21,Int 1)
(22,Int 1)
(23,Int 0)
(24,Int 4)
(25,Int 1)
(26,Int 4)
(27,Int 2)
(28,Int 2)
(29,Int 0)
(30,Int 3)
(31,Int 2)
(32,Int 5)
(33,Int 2)
(34,Int 3)
(35,Int 0)
(36,Int 1)
(37,Int 1)
(38,Int 5)
(39,Int 3)
(40,Int 3)
(41,Int 19)
(42,Int 3)
(43,Int 3)
(44,Bool True)
(45,Int 44)
(46,Bool True)
(47,Int 0)
(48,Int 0)
(49,Int (-1))
(50,Int 252)
(51,Int 1)
(52,Int 42)
(53,Int 46)
(54,Bool True)
(55,Int 46)
(56,Int 60)
(57,Int 233)
(58,Int 42)
(59,Int 39)
(60,Int 43)
(61,Int 3)
(62,Int 5)
(63,Int 47714)
(64,Int 0)
(65,Int 53)
(66,Int 70)
(67,Int 205)
(68,Int 37)
(69,Int 61)
(70,Int 38)
(71,Int 5)
(72,Int 65)
(73,Int 47714)
(74,Int 560651925)
(75,Int 65535)
(76,Null)
(77,Null)
(78,Null)
(79,Null)
(80,Null)
(81,Null)
(82,Null)
(83,Null)
(84,Null)
(85,Null)
(86,Null)
(87,Null)
(88,Null)
(89,Null)
(90,Null)
(91,Null)
(92,Null)
(93,Null)
(94,Null)
(95,Null)
(96,Null)
(97,Null)
(98,Null)
(99,Null)
